message/sequencer.go

```diff
@@ -1,92 +1,160 @@
 package message
 
 import (
+	"fmt"
 	"io"
 	"time"
 
-	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 	pb "go.gazette.dev/core/broker/protocol"
 )
 
 // Sequencer observes read-uncommitted messages from journals and sequences them
 // into acknowledged, read-committed messages. Read uncommitted messages are fed
 // to QueueUncommitted, after which the client must repeatedly call
-// DequeCommitted to drain all acknowledged messages until io.EOF is returned.
+// Step to dequeue all acknowledged messages until io.EOF is returned.
 //
 // In more detail, messages observed by QueueUncommitted may acknowledge one
 // or more pending messages previously observed by QueueUncommitted. For example,
 // a non-duplicate message with Flag_OUTSIDE_TXN acknowledges itself, and a
 // message with Flag_ACK_TXN also acknowledges messages having a lower clock.
-// DequeCommitted will drain the complete set of now-committed messages,
+// Step will drain the complete set of now-committed messages into field Dequeued,
 // and then return io.EOF.
 //
 // An advantage of the design is that no head-of-line blocking occurs: committed
-// messages are immediately deque'd upon observing their corresponding ACK_TXN,
+// messages are immediately dequeued upon observing their corresponding ACK_TXN,
 // even if they're interleaved with still-pending messages of other producers.
 //
 // Sequencer maintains an internal ring buffer of messages, which is usually
 // sufficient to directly read committed messages. When recovering from a
 // checkpoint, or if a very long sequence or old producer is acknowledged, it
 // may be necessary to start a replay of already-read messages. In this case:
-//  * DequeCommitted will return ErrMustStartReplay.
-//  * ReplayRange will return the exact offset range required.
+//  * QueueUncommitted will return QueueAckCommitReplay.
+//  * The client calls ReplayRange to determine the exact offset range required.
 //  * The client must then supply an appropriate Iterator to StartReplay.
-// Having done this, calls to DequeCommitted may resume to drain messages.
+// Having done this, calls to Step may resume to drain messages.
 type Sequencer struct {
+	// Dequeued is non-nil if (and only if) the Sequencer is in the
+	// process of dequeuing an acknowledged sequence of messages.
+	// After the client is done inspecting Dequeued, Step() must be
+	// called to step to the next Envelope of the sequence, or to
+	// complete the sequence (in which case Step returns io.EOF and
+	// Dequeued is now nil).
+	Dequeued *Envelope
+	// Clock of the current non-nil Dequeued, or zero.
+	// dequeuedClock is used to tighten the minimum clock bound once
+	// Step is called, indicating Dequeued has been processed.
+	// We could just ask Dequeued for its UUID, but the application
+	// may have changed it out from under us. That's especially easy
+	// to do if the application published it (which assigns a new UUID).
```

Comment:
The `Dequeued` message is passed to the application's `ConsumeMessage` callback, and that can do whatever it likes with the message. We can't _assume_ that its UUID continues to be the UUID we read. And in fact it won't be, if the application immediately published the message to another journal (as that sets a new UUID value on the decoded instance).

Reply:
It looks like [we're passing the `Envelope` to `ConsumeMessage` by value, rather than by reference](https://github.com/gazette/core/blob/75eedeeeb01bd1a85836f91760dec640d0d4ff5c/consumer/transaction.go#L305-L305), right? The `Application` shouldn't be able to mutate the message that's sitting in the `Sequencer`.

Even if we say that it could potentially mutate it and we need to defend against that here, phrasing it as "change the Application" seems a bit misleading to me. If you're republishing the same message to a different journal, it's just a different message now. This seems reflected in the reality of what's happening: the UUID of the message is changing. Is this the right way to think about this?

Reply:
The `Envelope` is passed by-value, including its contained `Message` field which is an interface (always a reference). We're essentially passing a reference by-value, but it's still a reference.